cpp/oneapi/dal/algo/kmeans/backend/cpu/train_kernel_lloyd_dense.cpp

```diff
@@ -0,0 +1,161 @@
+/*******************************************************************************
+* Copyright 2020 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include <daal/src/algorithms/kmeans/kmeans_init_kernel.h>
+#include <daal/src/algorithms/kmeans/kmeans_lloyd_kernel.h>
+
+#include "oneapi/dal/algo/kmeans/backend/cpu/train_kernel.hpp"
+#include "oneapi/dal/backend/interop/common.hpp"
+#include "oneapi/dal/backend/interop/table_conversion.hpp"
+#include "oneapi/dal/exceptions.hpp"
+
+namespace oneapi::dal::kmeans::backend {
+
+using std::int64_t;
+using dal::backend::context_cpu;
+
+namespace daal_kmeans      = daal::algorithms::kmeans;
+namespace daal_kmeans_init = daal::algorithms::kmeans::init;
+namespace interop          = dal::backend::interop;
+
+template <typename Float, daal::CpuType Cpu>
+using daal_kmeans_lloyd_dense_kernel_t =
+    daal_kmeans::internal::KMeansBatchKernel<daal_kmeans::lloydDense, Float, Cpu>;
+
+template <typename Float, daal::CpuType Cpu>
+using daal_kmeans_init_plus_plus_dense_kernel_t =
+    daal_kmeans_init::internal::KMeansInitKernel<daal_kmeans_init::plusPlusDense, Float, Cpu>;
+
+template <typename Float>
+static train_result call_daal_kernel(const context_cpu& ctx,
+                                     const descriptor_base& desc,
+                                     const table& data,
+                                     const table& initial_centroids) {
+    const int64_t row_count    = data.get_row_count();
+    const int64_t column_count = data.get_column_count();
+
+    const int64_t cluster_count       = desc.get_cluster_count();
+    const int64_t max_iteration_count = desc.get_max_iteration_count();
+    const double accuracy_threshold   = desc.get_accuracy_threshold();
+
+    daal_kmeans::Parameter par(cluster_count, max_iteration_count);
+    par.accuracyThreshold = accuracy_threshold;
+
+    auto arr_data        = row_accessor<const Float>{ data }.pull();
+    const auto daal_data = interop::convert_to_daal_homogen_table(arr_data,
+                                                                  data.get_row_count(),
+                                                                  data.get_column_count());
+
+    auto new_initial_centroids = initial_centroids;
+    if (!new_initial_centroids.has_data()) {
```

Comment:
Is this planned to be done as an algorithm inside an algorithm?

Reply:
Decided to postpone